### PR TITLE
Dumping files

### PR DESCRIPTION
**Pull request by Quaker762**
Added a function to dump virtual files to disk, so as to aid in reverse
engineering any file types that are currently unknown.

### CODE REVIEW

File: source/SH3/arc/vfile.cpp [hunk 1]
```diff
@@ -100,3 +102,23 @@ std::size_t sh3_arc_vfile::ReadData(void* destination, std::size_t len, read_err
...
+    {
+        Log(LogLevel::WARN, "sh3_arc_vfile::Dump2Disk( ): Warning! Attempting to flush unopen or empty buffer to disk!");
+        return;
+    }
+
+    // Firstly, we need to convert the file path to just the filename.
+    std::stringstream ss(fname);
+    std::string item;
+
+    while(getline(ss, item, '/')); // Throw out all strings except the last one (our file name)
```
Comment (z33ky) [inline source/SH3/arc/vfile.cpp]: It'd be cheaper to look for the last `/` in the string (see `std::string::rfind`) and passing an iterator pair to `std::ofstream`'s constructor (`begin(fname) + pos` and `end(fname)`).

File: source/SH3/arc/vfile.cpp [hunk 2]
```diff
@@ -100,3 +102,23 @@ std::size_t sh3_arc_vfile::ReadData(void* destination, std::size_t len, read_err
...
+
+    std::ofstream out_file(item);
+    for(int i = 0; i < buffer.size(); i++)
+        out_file << buffer[i];
```
Comment (z33ky) [inline source/SH3/arc/vfile.cpp]: The "streaming operators" (`<<`, `>>`) are meant for formatted data. We're dealing with binary data, so use `std::ofstream::write` instead. And you don't need to do it item by item then, you can just `write(buffer.data(), buffer.size())`.

File: source/SH3/arc/vfile.cpp [hunk 3]
```diff
@@ -100,3 +102,23 @@ std::size_t sh3_arc_vfile::ReadData(void* destination, std::size_t len, read_err
...
+        return;
+    }
+
+    // Firstly, we need to convert the file path to just the filename.
+    std::stringstream ss(fname);
+    std::string item;
+
+    while(getline(ss, item, '/')); // Throw out all strings except the last one (our file name)
+
+    std::ofstream out_file(item);
```
Comment (z33ky) [inline source/SH3/arc/vfile.cpp]: Check if opening fails.

File: source/SH3/arc/vfile.cpp [hunk 4]
```diff
@@ -100,3 +102,23 @@ std::size_t sh3_arc_vfile::ReadData(void* destination, std::size_t len, read_err
 
     return nbytes;
 }
+
+void sh3_arc_vfile::Dump2Disk()
+{
+    if(!open || buffer.empty())
```
Comment (z33ky) [inline source/SH3/arc/vfile.cpp:108]: Shouldn't be just create an empty file if `buffer.empty()`? Is this an error?

**Review by z33ky — CHANGES_REQUESTED** [review]

File: source/SH3/arc/vfile.cpp (hunk 4 above)
Comment (Quaker762) [reply source/SH3/arc/vfile.cpp:108]: Not so much an error, it just doesn't dump anything to the disk to save a few clock cycles. I didn't really see the point of creating an empty file, though I did initially have it in.

**Comment by Quaker762** [comment]
@z33ky  I think I've just noticed a problem with how I have this set up. Some mesages may have more than one attribute, but if we change `currentAttr` during `Load()`, the old attribute will be lost. How do you think we should store all the attributes??

**Comment by z33ky** [comment]
`currentAttr` is part of the other pull request #65. I'll answer there.
Could you add `msg.cpp` to the `source/CMakeLists.txt` as well?

**Comment by Quaker762** [comment]
Oops, I didn't even notice that. Is this PR ready for a merge?

> Could you add msg.cpp to the source/CMakeLists.txt as well?

I have in the last pushed commit in #67

**Comment by z33ky** [comment]
Doesn't look like it
```
$ git diff --name-only upstream/master quaker/dump
SILENT HILL 3 Redux.cbp
include/SH3/arc/vfile.hpp
include/SH3/system/msg/msg.hpp
source/SH3/arc/section.cpp
source/SH3/arc/types.cpp
source/SH3/arc/vfile.cpp
source/SH3/system/msg/msg.cpp
```

**Comment by Quaker762** [comment]
Oops, I didn't realise you meant to the dump branch, I'll do that and rebasr asap.

**Review by z33ky — APPROVED** [review]

**Comment by z33ky** [comment]
Well that is confusing. It seems you have some changes intended for #65 in here, as was also previously the case as seen by my `git diff` output before your rebase. You should remove those commits (including of course adding `msg.cpp` to `CMakeLists.txt`) from this pull request, especially since the message stuff is still subject to change.

**Comment by Quaker762** [comment]
Okay, I really fucked that rebase up badly (I must've pushed a button accidentally...). I'm working on fixing it now...

**Comment by Quaker762** [comment]
I'm deleting this branch and closing, I've fucked this up too hard.

**Pull request closed without merging**